packages/autoloader/src/class-autoloader.php

```diff
@@ -0,0 +1,140 @@
+<?php
+/* HEADER */ // phpcs:ignore
+
+/**
+ * This class handles management of the actual PHP autoloader.
+ */
+class Autoloader {
+
+	/**
+	 * Checks to see whether or not the autoloader should be initialized and then initializes it if so.
+	 *
+	 * @param Container|null $container The container we want to use for autoloader initialization. If none is given
+	 *                                  then a container will be created automatically.
+	 */
+	public static function init( $container = null ) {
+		// The container holds and manages the lifecycle of our dependencies
+		// to make them easier to work with and increase flexibility.
+		if ( ! isset( $container ) ) {
+			require_once __DIR__ . '/class-container.php';
+			$container = new Container();
+		}
+
+		// phpcs:disable Generic.Commenting.DocComment.MissingShort
+
+		/** @var Plugins_Handler $plugins_handler */
+		$plugins_handler = $container->get( Plugins_Handler::class );
+
+		// The cached plugins are all of those that were active or discovered by the autoloader during a previous request.
+		// Note that it's possible this list will include plugins that have since been deactivated, but after a request
+		// the cache should be updated and the deactivated plugins will be removed.
+		$cached_plugins = $plugins_handler->get_cached_plugins();
+
+		// The active plugins are those that we were able to discover in on the site. This list will not include
+		// mu-plugins, those activated by code, or those who are hidden by filtering.
+		// By combining these lists we can preemptively load classes for plugins that are activated via another method.
+		// While this may result in us considering packages in deactivated plugins the request after they're removed,
+		// there shouldn't be any problems as a result and the eventual consistency is reliable enough.
+		$all_plugins = array_values( array_unique( array_merge( $plugins_handler->get_active_plugins(), $cached_plugins ) ) );
+
+		/** @var Latest_Autoloader_Guard $guard */
+		$guard = $container->get( Latest_Autoloader_Guard::class );
+		if ( $guard->should_stop_init( $all_plugins ) ) {
+			return;
+		}
+
+		/** @var Autoloader_Handler $autoloader_handler */
+		$autoloader_handler = $container->get( Autoloader_Handler::class );
+
+		// Initialize the autoloader using the handler now that we're ready.
+		$autoloader_handler->create_autoloader( $all_plugins );
+
+		/** @var Hook_Manager $hook_manager */
+		$hook_manager = $container->get( Hook_Manager::class );
+
+		// When the active and cached plugin lists do not match we should
+		// update the cache. This will prevent plugins that have been
+		// deactivated from being considered in other requests.
+		$hook_manager->add_action(
+			'shutdown',
+			function () use ( $plugins_handler, $cached_plugins ) {
+				// Don't save a broken cache if an error happens during some plugin's initialization.
+				if ( ! did_action( 'plugins_loaded' ) ) {
+					return;
```

Comment:
Should the cache be cleared if an error happens during a plugin's initialization?

For example, a plugin could load a version of a package that conflicts with another plugin, and the problem plugin could be deactivated. However, it would still be in the cache and its packages could still be loaded, right?

Reply:
That seems like a good precaution to take. The goal here was to avoid cases where the cache would lead to further errors and that adds to it.